106-self-healing-feature-in-operator.md

```diff
@@ -0,0 +1,295 @@
+When it is created, the `KafkaRebalanceAssemblyOperator` doesn't run any rebalancing.
+This is because it doesn't represent an "actual" rebalance request to get an optimization proposal, but it's just the place where configuration related to auto-rebalancing is defined.
+The user can specify rebalancing goals and other configuration for rebalancing, within the resource.
+
+Here is an example template:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-full-rebalance-template
+  annotations:
+    strimzi.io/rebalance-template: "true" # specifies that this KafkaRebalance is a rebalance configuration template
+spec:
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
+  # ... other rebalancing related configuration
+```
+When the "template" is set the operator automatically creates (or updates) a corresponding "actual" `KafkaRebalance` custom resource based on the "template" when an anomaly is detected and notified by the `AnomalyDetectorNotifier`
+The operator copies over goals and rebalancing options from the referenced "template" resource to the "actual" rebalancing one and also adds the `spec.mode` to it.
+The "actual" `KafkaRebalance` custom resource will be named as `<my-cluster-name>-auto-rebalancing-full` where the `<my-cluster-name>` part comes from the `metadata.name` in the `Kafka` custom resource, and `full` refers to applying rebalance on all the brokers.
+
+#### AnomalyDetectorNotifier
+
+Cruise Control provides the `AnomalyNotifier` interface, which has multiple abstract methods on what to do if certain anomalies are detected.
+`SelfHealingNotifier` is the base class that contains the logic of self-healing and override the methods in by implementing the `AnomalyNotifier` interface.
+Some of those methods are:`onGoalViolation()`, `onBrokerFailure()`, `onDiskFailure`, `alert()` etc.
+The `AnomalyDetectorNotifier` will be based of the `SelfHealingNotifier` class.
+The anomalies with smaller priority value and detected time will be considered priority and resolved first.
+In case the anomaly is unfixable then they will be ignored.
+The `AnomalyDetectorNotifier` will override the `alert()` method and other detector methods of the `SelfHealingNotifier` to ensure that it is able to alert the operator whenever an anomaly is detected.
+Upon detection of an anomaly, the notifier would apply `strimzi.io/anomaly-detected: <anomaly-type>` annotation on the `Kafka` CR.
+Based on the annotation applied from this notifier, the operator will the run the rebalance on the brokers.
+
+The annotated Kafka resource would look like this:
+
+```yaml
+- apiVersion: kafka.strimzi.io/v1beta2
+  kind: Kafka
+  metadata:
+    annotations:
+      strimzi.io/anomaly-detected: "GOAL_VIOLATION"
```

Comment:
Shouldn't we have a little bit more complex structure (using a JSON?) also containing the anomalyId? It would be useful for debugging purposes I guess.